File: server/hall_server/client_service.ts
import express from "express";
import type { Request, Response } from "express";

import * as crypto from "../utils/crypto";
import * as db from "../utils/db";
import * as http from "../utils/http";
import * as room_service from "./room_service";

import type { HallServerConfig } from "../types/config";
import type { MessageRow, UserBriefRow } from "../types/db_rows";

var app = express();
// `config` 由 `start($config)` 在开始监听前赋值，所有请求处理里必然已经就绪；
// 下面的 `config!` 只是把这一点写出来，不改变任何运行时行为。
var config: HallServerConfig | null = null;

/** `req.query` 在本仓库里的真实形态：express 的类型放宽成了联合类型，这里在边界处收敛为字符串表。 */
type ClientQuery = http.QueryParams;

/** 调用方的 `req`，query 已按本仓库的约定收敛；`check_account` 收这个形状。 */
type ClientRequest = Request & { query: ClientQuery };

/** 客户端传来的 userId：`t_users.userid` 是 INT，但老代码一律按字符串拼签名与 query。 */
type UserId = string | number;

/**
 * `/login` 的返回体：先建对象、命中房间时再补 `roomid`（与迁移前逐字相同）。
 *
 * `sex` 是历史遗留：老代码读的是 `data.sex`，而 `get_user_data` 的 SELECT
 * （`SELECT userid,account,name,lv,exp,coins,gems,roomid`）里**没有 sex 列**，
 * 因此这个字段在运行时恒为 undefined，`JSON.stringify` 会直接把它丢掉。
 * 迁移保持原行为，所以这里声明成可选并如实标注。
 */
interface LoginEntry {
	roomid?: string;
	/** 客户端 IP（已剥掉 `::ffff:` 前缀）。 */
	ip: string;
	account: string;
	userid: number;
	name: string | null;
	lv: number;
	exp: number;
	coins: number;
	gems: number;
	sex?: number | null;
	[key: string]: unknown;
}

/** `/create_private_room`、`/enter_private_room` 的返回体；`sign` 是最后挂上去的进房签名。 */
interface RoomEntry {
	/** `/enter_private_room` 的 roomid 直接取自 query（`queryString` 可能给出 undefined）。 */
	roomid?: string;
	/** 与 room_service 的 `EnterInfo` 一致：这两项就是登记时的 query 字符串。 */
	ip?: string;
	port?: string;
	token: string;
	time: number;
	sign?: string;
	[key: string]: unknown;
}

/**
 * 校验 `account` / `sign` 两个 query 参数是否存在（签名校验本体历史上已被注释掉，保持原样）。
 *
 * @param req express 请求（query 已收敛为字符串表）。
 * @param res express 响应。
 * @returns 参数齐全时 true；否则已经回过错误响应并返回 false。
 */
function check_account(req: ClientRequest, res: Response): boolean {
	var account = req.query.account;
	var sign = req.query.sign;
	if(account == null || sign == null){
		http.send(res,1,"unknown error");
		return false;
	}
	/*
	var serverSign = crypto.md5(account + req.ip + config.ACCOUNT_PRI_KEY);
	if(serverSign != sign){
		http.send(res,2,"login failed.");
		return false;
	}
	*/
	return true;
}

//设置跨域访问
app.all('*', function(req: Request, res: Response, next) {
	res.header("Access-Control-Allow-Origin", "*");
	res.header("Access-Control-Allow-Headers", "X-Requested-With");
	res.header("Access-Control-Allow-Methods","PUT,POST,GET,DELETE,OPTIONS");
	res.header("X-Powered-By",' 3.2.1');
	res.header("Content-Type", "application/json;charset=utf-8");
	next();
});

app.get('/login',function(req: Request,res: Response){
	if(!check_account(req as ClientRequest,res)){
		return;
	}
	
	// @types/express 把 req.ip 标成 string | undefined（trust proxy 的边界情况）；
	// 这里与迁移前一样直接按字符串使用，不加存在性判断，行为逐字不变。
	var ip = req.ip as string;
	if(ip.indexOf("::ffff:") != -1){
		ip = ip.substr(7);
	}
	
	var account = http.queryString(req,"account");
	db.get_user_data(account,function(data: UserBriefRow | null){
		if(data == null){
			http.send(res,0,"ok");
			return;
		}

		var ret: LoginEntry = {
			account:data.account,
			userid:data.userid,
			name:data.name,
			lv:data.lv,
			exp:data.exp,
			coins:data.coins,
			gems:data.gems,
			ip:ip,
			// 历史行为：`sex` 不在 get_user_data 的 SELECT 里（见 LoginEntry 的说明），
			// 老代码直接读 `data.sex`，拿到的是 undefined。这里的断言只是让类型说得通，
			// 不做任何运行时判断或补默认值。
			sex:(data as { sex?: number | null }).sex,
		};

		db.get_room_id_of_user(data.userid,function(roomId: string | null){
			//如果用户处于房间中，则需要对其房间进行检查。 如果房间还在，则通知用户进入
			if(roomId != null){
				//检查房间是否存在于数据库中
				db.is_room_exist(roomId,function (retval: boolean){
					if(retval){
						ret.roomid = roomId;
					}
					else{
						//如果房间不在了，表示信息不同步，清除掉用户记录
						db.set_room_id_of_user(data.userid,null);
					}
					http.send(res,0,"ok",ret);
				});
			}
			else {
				http.send(res,0,"ok",ret);
			}
		});
	});
});

app.get('/create_user',function(req: Request,res: Response){
	if(!check_account(req as ClientRequest,res)){
		return;
	}
	var account = http.queryString(req,"account");
	var name = http.queryString(req,"name");
	var coins = 1000;
	var gems = 21;
	console.log(name);

	db.is_user_exist(account,function(ret: boolean){
		if(!ret){
			db.create_user(account,name,coins,gems,0,null,function(ret: boolean){
				if (ret == null) {
					http.send(res,2,"system error.");
				}
				else{
					http.send(res,0,"ok");					
				}
			});
		}
		else{
			http.send(res,1,"account have already exist.");
		}
	});
});

app.get('/create_private_room',function(req: Request,res: Response){
	//验证参数合法性
	var data = req.query as ClientQuery;
	//验证玩家身份
	if(!check_account(req as ClientRequest,res)){
		return;
	}

	var account = http.queryString(req,"account");

	// 保持迁移前的原样：这两个字段会被就地清掉（http.send 之后不回读它们）
	data.account = null;
	data.sign = null;
	var conf = data.conf;
	db.get_user_data(account,function(data: UserBriefRow | null){
		if(data == null){
			http.send(res,1,"system error");
			return;
		}
		var userId = data.userid;
		var name = data.name;
		//验证玩家状态
		db.get_room_id_of_user(userId,function(roomId: string | null){
			if(roomId != null){
				http.send(res,-1,"user is playing in room now.");
				return;
			}
			//创建房间
			room_service.createRoom(account,userId,conf,function(err: number,roomId: string | null){
				if(err == 0 && roomId != null){
					room_service.enterRoom(userId,name,roomId,function(errcode: number,enterInfo: room_service.EnterInfo | null){
						if(enterInfo){
							var ret: RoomEntry = {
								roomid:roomId,
								ip:enterInfo.ip,
								port:enterInfo.port,
								token:enterInfo.token,
								time:Date.now()
							};
							ret.sign = crypto.md5(ret.roomid + ret.token + ret.time + config!.ROOM_PRI_KEY);
							http.send(res,0,"ok",ret);
						}
						else{
							http.send(res,errcode,"room doesn't exist.");
						}
					});
				}
				else{
					http.send(res,err,"create failed.");					
				}
			});
		});
	});
});

app.get('/enter_private_room',function(req: Request,res: Response){
	var data = req.query as ClientQuery;
	var roomId = http.queryString(req,"roomid");
	if(roomId == null){
		http.send(res,-1,"parameters don't match api requirements.");
		return;
	}
	if(!check_account(req as ClientRequest,res)){
		return;
	}

	var account = data.account;

	db.get_user_data(account,function(data: UserBriefRow | null){
		if(data == null){
			http.send(res,-1,"system error");
			return;
		}
		var userId = data.userid;
		var name = data.name;

		//验证玩家状态
		//todo
		//进入房间
		room_service.enterRoom(userId,name,roomId,function(errcode: number,enterInfo: room_service.EnterInfo | null){
			if(enterInfo){
				var ret: RoomEntry = {
					roomid:roomId,
					ip:enterInfo.ip,
					port:enterInfo.port,
					token:enterInfo.token,
					time:Date.now()
				};
				ret.sign = crypto.md5(roomId + ret.token + ret.time + config!.ROOM_PRI_KEY);
				http.send(res,0,"ok",ret);
			}
			else{
				http.send(res,errcode,"enter room failed.");
			}
		});
	});
});

app.get('/get_history_list',function(req: Request,res: Response){
	var data = req.query as ClientQuery;
	if(!check_account(req as ClientRequest,res)){
		return;
	}
	var account = data.account;
	db.get_user_data(account,function(data: UserBriefRow | null){
		if(data == null){
			http.send(res,-1,"system error");
			return;
		}
		var userId = data.userid;
		db.get_user_history(userId,function(history: unknown[] | null){
			http.send(res,0,"ok",{history:history});
		});
	});
});

app.get('/get_games_of_room',function(req: Request,res: Response){
	var data = req.query as ClientQuery;
	var uuid = data.uuid;
	if(uuid == null){
		http.send(res,-1,"parameters don't match api requirements.");
		return;
	}
	if(!check_account(req as ClientRequest,res)){
		return;
	}
	db.get_games_of_room(uuid,function(data: unknown){
		console.log(data);
		http.send(res,0,"ok",{data:data});
	});
});

app.get('/get_detail_of_game',function(req: Request,res: Response){
	var data = req.query as ClientQuery;
	var uuid = data.uuid;
	var index = data.index;
	if(uuid == null || index == null){
		http.send(res,-1,"parameters don't match api requirements.");
		return;
	}
	if(!check_account(req as ClientRequest,res)){
		return;
	}
	db.get_detail_of_game(uuid,index,function(data: unknown){
		http.send(res,0,"ok",{data:data});
	});
});

app.get('/get_user_status',function(req: Request,res: Response){
	if(!check_account(req as ClientRequest,res)){
		return;
	}
	var account = http.queryString(req,"account");
	db.get_gems(account,function(data: { gems: number } | null){
		if(data != null){
			http.send(res,0,"ok",{gems:data.gems});	
		}
		else{
			http.send(res,1,"get gems failed.");
		}
	});
});

app.get('/get_message',function(req: Request,res: Response){
	if(!check_account(req as ClientRequest,res)){
		return;
	}
	var type = http.queryString(req,"type");
	
	if(type == null){
		http.send(res,-1,"parameters don't match api requirements.");
		return;
	}
	
	var version = http.queryString(req,"version");
	db.get_message(type,version,function(data: MessageRow | null){
		if(data != null){
			http.send(res,0,"ok",{msg:data.msg,version:data.version});	
		}
		else{
			http.send(res,1,"get message failed.");
		}
	});
});

export function start($config: HallServerConfig){
	config = $config;
	// 返回 http.Server，交给 app.js 汇总成启动横幅（见 utils/startup.js）
	return app.listen(config.CLEINT_PORT);
}

File: server-python/hall_server/client_service.py
"""大厅服客户端 HTTP 接口（:9001）。

对应 `server/hall_server/client_service.ts`。

`check_account()` 里的签名校验**历史上就被注释掉了**，只保留了"account / sign 两个参数
必须存在"的检查（保持原样，不要"顺手补回来"）。

进房签名（`/create_private_room`、`/enter_private_room`）必须与游戏服
`socket_service.py` 的登录校验逐字一致：
`md5(roomid + token + time + ROOM_PRI_KEY)`。

**封禁拦截**：`/login` 与三个建房 / 进房接口在动数据库之前先问一次
`utils.bancheck`（管理平台的内部只读接口）。被拦下时返回
`errcode = ERR_ACCOUNT_BANNED(3)`，客户端 `UserMgr.onLogin` 认这个码并把 `errmsg`
原样提示给玩家。问不到平台时是 fail-open 放行，理由见 `utils/bancheck.py`。
"""

from __future__ import annotations

import json
from typing import Any

from aiohttp import web

from hall_server import room_service
from utils import bancheck, crypto, db, http
from utils.jscompat import now_ms

#: 账号被封禁时的业务码（客户端 `UserMgr.onLogin` 认这个码弹提示）。
#: 1 是"参数不全"、2 是历史遗留的 login failed，3 空着，正好给封禁。
ERR_ACCOUNT_BANNED = 3

#: `config` 由 `create_routes()` 在开始监听前赋值，所有请求处理里必然已经就绪。
_config: dict[str, Any] | None = None


def _require_config() -> dict[str, Any]:
    if _config is None:
        raise RuntimeError("client_service.create_routes() 尚未调用")
    return _config


async def check_banned(account: str | None) -> web.Response | None:
    """账号被封禁时返回已经构造好的响应；没被封（或问不到平台）时返回 `None`。

    每个需要拦的接口在"取到 account 之后、查库之前"调它一次即可。

    :param account: 客户端带来的账号。
    """
    status = await bancheck.check_account(account)
    if not status.banned:
        return None
    return http.send(ERR_ACCOUNT_BANNED, bancheck.ban_message(status))


def check_account(request: web.Request) -> web.Response | None:
    """校验 `account` / `sign` 两个 query 参数是否存在。

    签名校验本体历史上已被注释掉（见模块文档），保持原样。

    :return: 参数齐全时 None；否则返回一个已经构造好的错误响应。
    """
    account = request.query.get("account")
    sign = request.query.get("sign")
    if account is None or sign is None:
        return http.send(1, "unknown error")
    # 下面这段在原实现里是被注释掉的：
    #   var serverSign = crypto.md5(account + req.ip + config.ACCOUNT_PRI_KEY);
    #   if(serverSign != sign){ http.send(res,2,"login failed."); return false; }
    return None


def create_routes(config: dict[str, Any]) -> web.Application:
    """建大厅服的客户端服务应用。"""
    global _config
    _config = config

    app = web.Application()

    @web.middleware
    async def cors(request: web.Request, handler: Any) -> web.StreamResponse:
        response = await handler(request)
        response.headers["Access-Control-Allow-Origin"] = "*"
        response.headers["Access-Control-Allow-Headers"] = "X-Requested-With"
        response.headers["Access-Control-Allow-Methods"] = "PUT,POST,GET,DELETE,OPTIONS"
        response.headers["X-Powered-By"] = " 3.2.1"
        if "Content-Type" not in response.headers:
            response.headers["Content-Type"] = "application/json;charset=utf-8"
        return response

    app.middlewares.append(cors)

    async def on_login(request: web.Request) -> web.Response:
        failed = check_account(request)
        if failed is not None:
            return failed

        ip = http.request_ip(request)
        if "::ffff:" in ip:
            ip = ip[7:]

        account = http.query_string(request, "account")
        # 封禁拦截：在查库之前问平台，被拦下就到此为止（客户端按 errcode=3 提示）。
        banned = await check_banned(account)
        if banned is not None:
            return banned

        data = await db.get_user_data(account)
        if data is None:
            return http.send(0, "ok")

        # 历史行为：`sex` 不在 get_user_data 的 SELECT 里，老代码直接读 `data.sex`
        # 拿到的是 undefined，`JSON.stringify` 会把这个键整个丢掉。
        # 这里同样"有值才放"，保证返回结构与 Node 版逐字一致。
        ret: dict[str, Any] = {
            "account": data.get("account"),
            "userid": data.get("userid"),
            "name": data.get("name"),
            "lv": data.get("lv"),
            "exp": data.get("exp"),
            "coins": data.get("coins"),
            "gems": data.get("gems"),
            "ip": ip,
        }
        sex = data.get("sex")
        if sex is not None:
            ret["sex"] = sex

        user_id = data.get("userid")
        room_id = await db.get_room_id_of_user(user_id)
        # 如果用户处于房间中，则需要对其房间进行检查。如果房间还在，则通知用户进入
        if room_id is not None:
            if await db.is_room_exist(room_id):
                ret["roomid"] = room_id
            else:
                # 如果房间不在了，表示信息不同步，清除掉用户记录
                await db.set_room_id_of_user(user_id, None)
        return http.send(0, "ok", ret)

    async def on_create_user(request: web.Request) -> web.Response:
        failed = check_account(request)
        if failed is not None:
            return failed
        account = http.query_string(request, "account")
        name = http.query_string(request, "name")
        coins = 1000
        gems = 21
        print(name)

        if not await db.is_user_exist(account):
            result = await db.create_user(account, name, coins, gems, 0, None)
            if result is None:
                return http.send(2, "system error.")
            return http.send(0, "ok")
        return http.send(1, "account have already exist.")

    async def on_create_private_room(request: web.Request) -> web.Response:
        # 保持迁移前的原样：这两个字段会被就地清掉（http.send 之后不回读它们）
        data = request.query
        failed = check_account(request)
        if failed is not None:
            return failed

        account = http.query_string(request, "account")
        conf = data.get("conf")
        _ = data.get("account"), data.get("sign")  # 原实现里的 `data.account = null; data.sign = null;`

        banned = await check_banned(account)
        if banned is not None:
            return banned

        user = await db.get_user_data(account)
        if user is None:
            return http.send(1, "system error")

        user_id = user.get("userid")
        name = user.get("name")
        room_id = await db.get_room_id_of_user(user_id)
        if room_id is not None:
            return http.send(-1, "user is playing in room now.")

        err, room_id = await room_service.create_room(account, user_id, conf)
        if err != 0 or room_id is None:
            return http.send(err, "create failed.")

        errcode, enter_info = await room_service.enter_room(user_id, name, room_id)
        if enter_info is not None:
            ret: dict[str, Any] = {
                "roomid": room_id,
                "ip": enter_info.ip,
                "port": enter_info.port,
                "token": enter_info.token,
                "time": now_ms(),
            }
            ret["sign"] = crypto.md5(
                str(ret["roomid"]) + str(ret["token"]) + str(ret["time"]) + _require_config()["ROOM_PRI_KEY"]
            )
            return http.send(0, "ok", ret)
        return http.send(errcode, "room doesn't exist.")

    async def on_create_single_room(request: web.Request) -> web.Response:
        """单人模式建房：与 `/create_private_room` 相同，但强制带上 `single:1`。

        差别只有两处：

        1. 客户端传来的 conf 里补上 `single:1`，其余选项照用——玩法（血流成河 / 血战到底）、
           底分、番数还是玩家在 CreateRoom 面板里选的那一套；
        2. conf 重新序列化后再交给游戏服。大厅服与游戏服各自算一次 md5，用的必须是
           **同一个字符串**，这里传下去的就是那一个。

        游戏服收到 `single` 后会预置三个机器人、并跳过房卡校验（见 `roommgr.create_room`）。
        """
        failed = check_account(request)
        if failed is not None:
            return failed

        account = http.query_string(request, "account")
        conf_raw = request.query.get("conf")
        if conf_raw is None:
            return http.send(-1, "parameters don't match api requirements.")
        try:
            conf = json.loads(conf_raw)
        except ValueError:
            return http.send(-1, "invalid conf.")
        if not isinstance(conf, dict):
            return http.send(-1, "invalid conf.")
        conf["single"] = 1
        conf_str = json.dumps(conf, separators=(",", ":"))

        # 单人模式也拦：封禁是账号级状态，不该因为"只跟机器人打"就绕过去。
        banned = await check_banned(account)
        if banned is not None:
            return banned

        user = await db.get_user_data(account)
        if user is None:
            return http.send(1, "system error")

        user_id = user.get("userid")
        name = user.get("name")
        room_id = await db.get_room_id_of_user(user_id)
        if room_id is not None:
            return http.send(-1, "user is playing in room now.")

        err, room_id = await room_service.create_room(account, user_id, conf_str)
        if err != 0 or room_id is None:
            return http.send(err, "create failed.")

        errcode, enter_info = await room_service.enter_room(user_id, name, room_id)
        if enter_info is not None:
            ret: dict[str, Any] = {
                "roomid": room_id,
                "ip": enter_info.ip,
                "port": enter_info.port,
                "token": enter_info.token,
                "time": now_ms(),
            }
            ret["sign"] = crypto.md5(
                str(ret["roomid"])
                + str(ret["token"])
                + str(ret["time"])
                + _require_config()["ROOM_PRI_KEY"]
            )
            return http.send(0, "ok", ret)
        return http.send(errcode, "room doesn't exist.")

    async def on_enter_private_room(request: web.Request) -> web.Response:
        room_id = http.query_string(request, "roomid")
        if room_id is None:
            return http.send(-1, "parameters don't match api requirements.")
        failed = check_account(request)
        if failed is not None:
            return failed

        account = request.query.get("account")
        banned = await check_banned(account)
        if banned is not None:
            return banned

        user = await db.get_user_data(account)
        if user is None:
            return http.send(-1, "system error")

        user_id = user.get("userid")
        name = user.get("name")

        # 验证玩家状态
        # todo
        # 进入房间
        errcode, enter_info = await room_service.enter_room(user_id, name, room_id)
        if enter_info is not None:
            ret: dict[str, Any] = {
                "roomid": room_id,
                "ip": enter_info.ip,
                "port": enter_info.port,
                "token": enter_info.token,
                "time": now_ms(),
            }
            ret["sign"] = crypto.md5(
                str(room_id) + str(ret["token"]) + str(ret["time"]) + _require_config()["ROOM_PRI_KEY"]
            )
            return http.send(0, "ok", ret)
        return http.send(errcode, "enter room failed.")

    async def on_get_history_list(request: web.Request) -> web.Response:
        failed = check_account(request)
        if failed is not None:
            return failed
        account = request.query.get("account")
        user = await db.get_user_data(account)
        if user is None:
            return http.send(-1, "system error")
        history = await db.get_user_history(user.get("userid"))
        return http.send(0, "ok", {"history": history})

    async def on_get_games_of_room(request: web.Request) -> web.Response:
        uuid = request.query.get("uuid")
        if uuid is None:
            return http.send(-1, "parameters don't match api requirements.")
        failed = check_account(request)
        if failed is not None:
            return failed
        data = await db.get_games_of_room(uuid)
        print(data)
        return http.send(0, "ok", {"data": data})

    async def on_get_detail_of_game(request: web.Request) -> web.Response:
        uuid = request.query.get("uuid")
        index = request.query.get("index")
        if uuid is None or index is None:
            return http.send(-1, "parameters don't match api requirements.")
        failed = check_account(request)
        if failed is not None:
            return failed
        data = await db.get_detail_of_game(uuid, index)
        return http.send(0, "ok", {"data": data})

    async def on_get_user_status(request: web.Request) -> web.Response:
        failed = check_account(request)
        if failed is not None:
            return failed
        account = http.query_string(request, "account")
        data = await db.get_gems(account)
        if data is not None:
            return http.send(0, "ok", {"gems": data.get("gems")})
        return http.send(1, "get gems failed.")

    async def on_get_message(request: web.Request) -> web.Response:
        failed = check_account(request)
        if failed is not None:
            return failed
        type_ = http.query_string(request, "type")
        if type_ is None:
            return http.send(-1, "parameters don't match api requirements.")
        version = http.query_string(request, "version")
        data = await db.get_message(type_, version)
        if data is not None:
            return http.send(0, "ok", {"msg": data.get("msg"), "version": data.get("version")})
        return http.send(1, "get message failed.")

    app.router.add_get("/login", on_login)
    app.router.add_get("/create_user", on_create_user)
    app.router.add_get("/create_private_room", on_create_private_room)
    app.router.add_get("/create_single_room", on_create_single_room)
    app.router.add_get("/enter_private_room", on_enter_private_room)
    app.router.add_get("/get_history_list", on_get_history_list)
    app.router.add_get("/get_games_of_room", on_get_games_of_room)
    app.router.add_get("/get_detail_of_game", on_get_detail_of_game)
    app.router.add_get("/get_user_status", on_get_user_status)
    app.router.add_get("/get_message", on_get_message)
    return app
